opengovsg/GoGovSG
# feat: admin api v1 - create url

## Problem

This PR closes [APIs for Forms to create shortlinks](https://www.notion.so/opengov/APIs-for-Forms-to-create-shortlinks-41bf2f8c7627441796677ba95edd8a63)

## Solution

This PR implements a new endpoint `{{host}}/api/v1/admin/urls` which is allows for whitelisted emails defined in `ADMIN_API_EMAILS` to call the endpoint to create new short URLs and transfer its ownership to a provided email address.

**Features**:

- Added new route to `{{host}}/api/v1/admin/urls` that uses `apiKeyAdminAuthMiddleware` for authentication
  - Updated `apiKeyAdminAuthMiddleware` to throw `jsonMessage` if api key is unauthorized
  - Added POST route that additionally validates the required `email` attribute on top of existing validations

- Added `AdminApiV1Controller` that handles POST requests with `shortUrl`, `longUrl`, and `email` attributes in the JSON body 
  - `createUrl` method to find or create a new user with `email` provided, then creates a new link with the authenticated user before transferring ownership to the parsed `email`.

**Improvements**:

- Renamed `ADMIN_API_EMAIL` env variable to `ADMIN_API_EMAILS` to reflect its use to white list multiple emails
  - Added `integration-test-admin@open.gov.sg` to the white list for integration tests

- Updated `isAdmin` method in `ApiKeyAuthService` to handle multiple white listed emails


## Tests
Unit Test Cases
- Create, sanitize and transfer link to target email for admin API
- Create and sanitize link with same owner and target email for admin API
- Reports bad request with user creation
- Reports not found on NotFoundError
- Reports bad request on AlreadyExistsError
- Reports bad request on Sequelize.ValidationError
- Reports bad request on generic error

Integration Test
- Added `createIntegrationTestAdminUser` to create a new integration test admin user which is using the defined `integration-test-admin@open.gov.sg` email
- Tests Cases
  - Should not be able to create urls without API key header
  - Should not be able to create urls with invalid API key
  - Should not be able to create urls with unauthorized API key
  - Should be able to create link url with longUrl, shortUrl, and validEmail
  - Should not be able to create link url with longUrl and shortUrl, without validEmail
  - Should not be able to create link url with longUrl and shortUrl, with invalid email
  - Should be able to create link url with longUrl and validEmail, without shortUrl
  - Should not be able to create link url with longUrl, without shortUrl and validEmail
  - Should not be able to create link url with invalid email
  - Should not be able to create link url without longUrl, shortUrl, and email
  - Should not be able to create link url with invalid longUrl
  - Should not be able to create link url with invalid shortUrl
  - Should be able to create link url with longUrl, shortUrl, and same email as admin
